[ERROR] libmpi_mpifh.so: undefined reference to `memcpy@GLIBC_2.14'

I use the conda to install the tdep code, which is obtained by git clone https://github.com/tdep-developers/tdep.git. However, i meet these following errors 

Building dump_dynamical_matrices
/home/cqwl222/anaconda3/envs/tdep/bin/../lib/gcc/x86_64-conda-linux-gnu/11.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: //home/cqwl222/anaconda3/envs/tdep/lib/./libpmix.so.2: undefined reference to `clock_gettime@GLIBC_2.17'
/home/cqwl222/anaconda3/envs/tdep/bin/../lib/gcc/x86_64-conda-linux-gnu/11.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: //home/cqwl222/anaconda3/envs/tdep/lib/libmpi_mpifh.so: undefined reference to `memcpy@GLIBC_2.14'
collect2: error: ld returned 1 exit status
make: *** [Makefile:24: ../../build/dump_dynamical_matrices/dump_dynamical_matrices] Error 1
  
How to resolve this problem.

Hello,

I think this has nothing to do with TDEP. Your anaconda environment is inconsistent

/home/cqwl222/anaconda3/envs/tdep/lib/./libpmix.so.2: undefined reference to clock_gettime@GLIBC_2.17' 
/home/cqwl222/anaconda3/envs/tdep/lib/libmpi_mpifh.so: undefined reference to memcpy@GLIBC_2.14'

These 2 libs have been installed with different glibc versions, so it's normal they don't link together properly. Try to re-install your conda stack.

It could be that the older releases did not use clock_gettime or memcpy in the same way, or something like that.

I download the code from branch (https://github.com/tdep-developers/tdep/tree/70-error-libmpi_mpifhso-undefined-reference-to-memcpyglibc_214)

/home/cqwl222/VASPTOOL/TDEP/conda_env/bin/../lib/gcc/x86_64-conda-linux-gnu/9.5.0/../../../../x86_64-conda-linux-gnu/bin/ld: //home/cqwl222/VASPTOOL/TDEP/conda_env/lib/libgcc_s.so.1: undefined reference to `memcpy@GLIBC_2.14'
collect2: error: ld returned 1 exit status
make: *** [Makefile:24: ../../build/dump_dynamical_matrices/dump_dynamical_matrices] Error 1

(/home/cqwl222/VASPTOOL/TDEP/conda_env) cqwl222@cqwl222:~/VASPTOOL/TDEP/tdep-70/src/dump_dynamical_matrices$ make clean
rm -f ../../build/dump_dynamical_matrices/dump_dynamical_matrices ../../build/dump_dynamical_matrices/main.o ../../build/dump_dynamical_matrices/options.o *.mod ../../build/dump_dynamical_matrices/*.mod
(/home/cqwl222/VASPTOOL/TDEP/conda_env) cqwl222@cqwl222:~/VASPTOOL/TDEP/tdep-70/src/dump_dynamical_matrices$ make
/home/cqwl222/VASPTOOL/TDEP/conda_env/bin/mpifort -ffree-line-length-none -std=f2008 -cpp -Wno-argument-mismatch  -DGPqt -Dclusterprogressbar -I../../inc/libolle -I../../inc/libflap -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -J ../../build/dump_dynamical_matrices/  -O0  -lolle -lflap -llapack -lscalapack -lhdf5 -lhdf5_fortran -lmpi_mpifh -lmpi -lfftw3 -c options.f90 -o ../../build/dump_dynamical_matrices/options.o
/home/cqwl222/VASPTOOL/TDEP/conda_env/bin/mpifort -ffree-line-length-none -std=f2008 -cpp -Wno-argument-mismatch  -DGPqt -Dclusterprogressbar -I../../inc/libolle -I../../inc/libflap -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -J ../../build/dump_dynamical_matrices/  -O0  -lolle -lflap -llapack -lscalapack -lhdf5 -lhdf5_fortran -lmpi_mpifh -lmpi -lfftw3 -c main.f90  -o ../../build/dump_dynamical_matrices/main.o
/home/cqwl222/VASPTOOL/TDEP/conda_env/bin/mpifort -ffree-line-length-none -std=f2008 -cpp -Wno-argument-mismatch  -DGPqt -Dclusterprogressbar -I../../inc/libolle -I../../inc/libflap -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -I//home/cqwl222/VASPTOOL/TDEP/conda_env/include -J ../../build/dump_dynamical_matrices/  -o ../../build/dump_dynamical_matrices/dump_dynamical_matrices ../../build/dump_dynamical_matrices/main.o ../../build/dump_dynamical_matrices/options.o -L../../lib -L//home/cqwl222/VASPTOOL/TDEP/conda_env/lib -L//home/cqwl222/VASPTOOL/TDEP/conda_env/lib -L//home/cqwl222/VASPTOOL/TDEP/conda_env/lib -L//home/cqwl222/VASPTOOL/TDEP/conda_env/lib -lolle -lflap -llapack -lscalapack -lhdf5 -lhdf5_fortran -lmpi_mpifh -lmpi -lfftw3
/home/cqwl222/VASPTOOL/TDEP/conda_env/bin/../lib/gcc/x86_64-conda-linux-gnu/9.5.0/../../../../x86_64-conda-linux-gnu/bin/ld: //home/cqwl222/VASPTOOL/TDEP/conda_env/lib/libgcc_s.so.1: undefined reference to `memcpy@GLIBC_2.14'
collect2: error: ld returned 1 exit status
make: *** [Makefile:24: ../../build/dump_dynamical_matrices/dump_dynamical_matrices] Error 1
(/home/cqwl222/VASPTOOL/TDEP/conda_env) cqwl222@cqwl222:~/VASPTOOL/TDEP/tdep-70/src/dump_dynamical_matrices$

--- a/src/dump_dynamical_matrices/Makefile
+++ b/src/dump_dynamical_matrices/Makefile
@@ -22,8 +22,7 @@ clean:
 	rm -f $(PROG) $(OBJS) *.mod $(OBJECT_PATH)*.mod
 
 
-$(OBJECT_PATH)main.o: $(OBJECT_PATH)options.o 
-	$(F90) $(F90FLAGS) -c main.f90 $(LIBS) -o $@
+$(OBJECT_PATH)main.o: $(OBJECT_PATH)options.o
+	$(F90) $(F90FLAGS) -c main.f90 -o $@
 $(OBJECT_PATH)options.o:
-	$(F90) $(F90FLAGS) -c options.f90 $(LIBS) -o $@
-
+	$(F90) $(F90FLAGS) -c options.f90 -o $@

--- a/src/dump_dynamical_matrices/Makefile
+++ b/src/dump_dynamical_matrices/Makefile
@@ -5,30 +5,25 @@ OBJECT_PATH=../../build/$(CODE)/
 
 OBJS = $(OBJECT_PATH)main.o $(OBJECT_PATH)options.o
 
-
-# some paths and stuff. Ok.
-IPATH = -I../../inc/libolle -I../../inc/libflap $(blaslapackIPATH) $(incIPATHhdf) $(incIPATHmpi) $(incIPATHfft)
-
 LPATH = -L../../lib $(blaslapackLPATH) $(incLPATHhdf) $(incLPATHmpi) $(incLPATHfft)
+IPATH = -I../../inc/libolle -I../../inc/libflap $(blaslapackIPATH) $(incIPATHhdf) $(incIPATHmpi) $(incIPATHfft)
 LIBS = -lolle -lflap $(blaslapackLIBS) $(incLIBShdf) $(incLIBSmpi) $(incLIBSfft)
 
-# ok, I think I get this.
-#OPT = -O0 -fbacktrace -fcheck=all -finit-real=nan -finit-derived -fmax-errors=10
-F90 = $(FC) $(IPATH) $(MODULE_FLAG) $(OBJECT_PATH) #$(warnings_gcc)
+#OPT = -O0 -fbacktrace -fcheck=all -Wall
+F90 = $(FC) $(LPATH) $(IPATH) $(MODULE_FLAG) $(OBJECT_PATH)
 F90FLAGS = $(OPT) $(MODS) $(LIBS)
 
-# I guess this is the normal one
 all: $(PROG)
 
 $(PROG): $(OBJS)
-	$(F90) -o $@ $(OBJS) $(LPATH) $(LIBS)
+	$(F90) $(LDFLAGS) -o $@ $(OBJS) $(LIBS)
 
 clean:
 	rm -f $(PROG) $(OBJS) *.mod $(OBJECT_PATH)*.mod
 
 
 $(OBJECT_PATH)main.o: $(OBJECT_PATH)options.o 
-	$(F90) $(F90FLAGS) -c main.f90  -o $@
+	$(F90) $(F90FLAGS) -c main.f90 $(LIBS) -o $@
 $(OBJECT_PATH)options.o:
-	$(F90) $(F90FLAGS) -c options.f90 -o $@
+	$(F90) $(F90FLAGS) -c options.f90 $(LIBS) -o $@
 

